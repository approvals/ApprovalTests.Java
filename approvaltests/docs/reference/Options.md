<a id="top"></a>

# Options

<!-- toc -->
## Contents

  * [Introduction](#introduction)
  * [Fluent Interface](#fluent-interface)
  * [Reporters](#reporters)
  * [Scrubbers](#scrubbers)
  * [File Options](#file-options)
    * [File Extensions](#file-extensions)
  * [Custom Comparators](#custom-comparators)
  * [Defaults](#defaults)
  * [Adding to Existing Options object](#adding-to-existing-options-object)<!-- endToc -->

## Introduction

There are many things you might want to tweak with Approval Tests. `Options` is the entry-point for many of the changes.
It is on all `verify()` methods, as an optional parameter.

## Fluent Interface

`Options` utilizes a fluent interface, allowing you to chain together commands. Each returned object is a new copy.

<!-- snippet: specify_all_the_options -->
<a id='snippet-specify_all_the_options'></a>
```java
new Options().withReporter(new ReportNothing()).withScrubber(new GuidScrubber()).forFile()
    .withExtension(".json");
```
<sup><a href='/approvaltests-tests/src/test/java/org/approvaltests/OptionsSamplesTest.java#L15-L18' title='Snippet source file'>snippet source</a> | <a href='#snippet-specify_all_the_options' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## Reporters

[Reporters](/docs/reference/Reporters.md#top) launch diff tools upon failure.  Read how to configure them with `Options` [here](Reporters.md#via-options):

## Scrubbers

[Scrubbers](/docs/explanations/Scrubbers.md#top) clean output to help remove inconsistent pieces of text, such as dates.  Read how to set up Scrubbers [here](../Scrubbers.md#configuring-scrubbers):
There are two ways to set a Scrubber.

## File Options

The `Options.forFile()` class exists to customise the `.approved` and `.received` files in various ways.

### File Extensions

If you want to change the file extension of both the approved and received files, use `withExtension()`.

<!-- snippet: basic_approval_with_file_extension -->
<a id='snippet-basic_approval_with_file_extension'></a>
```java
Approvals.verify("text to be verified", new Options().forFile().withExtension(".xyz"));
```
<sup><a href='/approvaltests-tests/src/test/java/org/approvaltests/OptionsSamplesTest.java#L23-L25' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic_approval_with_file_extension' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

**Note:** `withExtension()` returns an `Options` object, so it's possible to keep appending more `with...()` calls.

## Custom Comparators
By default Approval Tests will only check if the `.approved` and `.received` files are _exactly_ matching.
The only accomodations it makes is for differences in line endings.

If you would like to create a more flexible comparison, you can do it by using the `Options.withComparator()` function.

## Defaults

The default constructor for Options does:

* no scrubbing
* uses file extension `.txt`
* uses whatever is currently set as the default reporter.

## Adding to Existing Options object

Each instance of Options is not mutable. Therefore, every call produces a new copy of the Options object. This allows you to modify a specific option from an existing Options object,
while retaining all other settings, and not changing the original object.


---

[Back to User Guide](../README.md#top)
