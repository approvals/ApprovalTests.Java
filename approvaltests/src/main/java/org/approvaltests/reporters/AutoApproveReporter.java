package org.approvaltests.reporters;

import com.spun.util.io.FileUtils;
import org.approvaltests.core.VerifyResult;

import java.io.File;

public class AutoApproveReporter implements ReporterWithApprovalPower
{
  @Override
  public boolean report(String received, String approved)
  {
    File a = new File(approved);
    if (a.exists())
    {
      a.delete();
    }
    File r = new File(received);
    FileUtils.copyFile(r, a);
    return true;
  }

  @Override
  public VerifyResult approveWhenReported()
  {
    return VerifyResult.SUCCESS;
  }
}
