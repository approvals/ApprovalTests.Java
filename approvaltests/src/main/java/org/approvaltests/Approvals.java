package org.approvaltests;

import com.spun.util.ArrayUtils;
import com.spun.util.FormattedException;
import com.spun.util.JsonUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.persistence.ExecutableCommand;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.*;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.MasterDirectoryNamer;
import org.approvaltests.namer.NamerFactoryForOptions;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class Approvals
{
  public static final NamerFactoryForOptions NAMES        = NamerFactoryForOptions.INSTANCE;
  public static Loader<ApprovalNamer>        namerCreater = new Loader<ApprovalNamer>()
                                                          {
                                                            public ApprovalNamer load()
                                                            {
                                                              return new StackTraceNamer();
                                                            }
                                                          };
  public static void verify(String response)
  {
    verify(response, new Options());
  }
  public static void verify(String response, Options options)
  {
    verify(options.createWriter(response), options);
  }
  public static void verify(Object object)
  {
    verify(object, new Options());
  }
  public static void verify(Object object, Options options)
  {
    checkForAwtComponents(object == null ? Object.class : object.getClass());
    verify(Objects.toString(object), options);
  }
  public static void verify(Verifiable object)
  {
    verify(object, new Options());
  }
  public static void verify(Verifiable object, Options options)
  {
    VerifyParameters verifyParameters = object.getVerifyParameters(options);
    verify(Objects.toString(object), verifyParameters.getOptions());
  }
  private static void checkForAwtComponents(Class<?> type)
  {
    if (type == Object.class)
    { return; }
    if ("java.awt.Component".equals(type.getCanonicalName()))
    {
      throw new FormattedException(
          "Approvals.verify(Component) has moved to AwtApprovals.verify(Component)\ncalled Approvals.verify(Object)");
    }
    checkForAwtComponents(type.getSuperclass());
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, T[] array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, T[] array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1, Options options)
  {
    verifyAll("", values, f1, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1, Options options)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(text, options);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1, new Options());
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1, Options options)
  {
    verifyAll(header, Arrays.asList(values), f1, options);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    verifyAll(header, array, f1, new Options());
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1, Options options)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(text, options);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, Iterable<T> array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static void verifyHtml(String response)
  {
    verifyHtml(response, new Options());
  }
  public static void verifyHtml(String response, Options options)
  {
    verify(response, options.forFile().withExtension(".html"));
  }
  public static void verify(File generateFile)
  {
    verify(generateFile, new Options());
  }
  public static void verify(File generateFile, Options options)
  {
    verify(options.createWriter(generateFile), options);
  }
  /**
   * @deprecated Use {@link #verify(ApprovalWriter, ApprovalNamer, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), new Options(reporter));
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer)
  {
    verify(writer, namer, new Options());
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, Options options)
  {
    verify(new FileApprover(writer, namer, options.getComparator()), options);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, new Options());
  }
  public static void verify(ApprovalWriter writer, Options options)
  {
    verify(writer, options.forFile().getNamer(), options);
  }
  public static void verifyXml(String xml)
  {
    verifyXml(xml, new Options());
  }
  public static void verifyXml(String xml, Options options)
  {
    final String formattedXml = ApprovalXmlWriter.prettyPrint(xml, 2);
    verify(formattedXml, options.forFile().withExtension(".xml"));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, new Options());
  }
  /**
   * @deprecated Use {@link #verify(ApprovalApprover, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, new Options(reporter));
  }
  public static void verify(ApprovalApprover approver, Options options)
  {
    ApprovalFailureReporter reporter = options.getReporter();
    VerifyResult result = approver.approve();
    if (result.isFailure())
    {
      result = approver.reportFailure(reporter);
    }
    if (result.isSuccessful())
    {
      approver.cleanUpAfterSuccess(reporter);
    }
    else
    {
      approver.fail();
    }
  }
  public static void verify(ExecutableCommand query)
  {
    verify(query, new Options());
  }
  public static void verify(ExecutableCommand query, Options options)
  {
    verify(query.getCommand(), ExecutableQueryFailure.create(query, options));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(map, new Options());
  }
  public static void verify(Map<?, ?> map, Options options)
  {
    verify(StringUtils.toString(map), options);
  }
  public static void verify(ResultSet rs)
  {
    verify(rs, new Options());
  }
  public static void verify(ResultSet rs, Options options)
  {
    verify(options.createWriter(rs), options);
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(loader, new Options());
  }
  public static <T> void verify(SqlLoader<T> loader, Options options)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader), options);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileInDirectory(directory, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(), options);
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files, Options options)
  {
    List<File> mismatched = new ArrayList<>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(options.createWriter(f), new MasterDirectoryNamer(f, options), options);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json)
  {
    JsonApprovals.verifyJson(json);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json, Options options)
  {
    JsonApprovals.verifyJson(json, options);
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(actual, fileExtensionWithoutDot, new Options());
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot, Options options)
  {
    verify("" + actual, options.forFile().withExtension(fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o)
  {
    JsonApprovals.verifyAsJson(o);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass)
  {
    verifyAsJson(o, gsonBuilder, gsonBuilderClass, new Options());
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass,
      Options options)
  {
    verify(JsonUtils.asJsonWithBuilder(o, gsonBuilder, gsonBuilderClass),
        options.forFile().withExtension(".json"));
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o, Options options)
  {
    JsonApprovals.verifyAsJson(o, options);
  }
  public static void verifyException(Action0 runnableBlock)
  {
    verifyException(runnableBlock, new Options());
  }
  public static void verifyException(Action0 runnableBlock, Options options)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null)
    { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()), options);
  }
  public static ApprovalSettings settings()
  {
    return new ApprovalSettings();
  }
}
