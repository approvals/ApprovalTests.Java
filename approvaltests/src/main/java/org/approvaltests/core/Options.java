package org.approvaltests.core;

import com.spun.util.ArrayUtils;
import org.approvaltests.Approvals;
import org.approvaltests.ReporterFactory;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.inline.InlineComparator;
import org.approvaltests.inline.InlineOptions;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.NamerWrapper;
import org.approvaltests.reporters.MultiReporter;
import org.approvaltests.scrubbers.NoOpScrubber;
import org.approvaltests.writers.ApprovalWriterFactory;
import org.approvaltests.writers.DefaultApprovalWriterFactory;
import org.lambda.functions.Function1;
import org.lambda.functions.Function2;

import java.io.File;
import java.util.EnumMap;
import java.util.Map;

public class Options
{
  private enum Fields {
                       SCRUBBER, REPORTER, FILE_OPTIONS_FILE_EXTENSION, FILE_OPTIONS_NAMER, WRITER, COMPARATOR;
  }
  private final Map<Fields, Object> fields = new EnumMap<>(Fields.class);
  public Options()
  {
  }

  public Options and(Function1<Options, Options> optionsUpdate)
  {
    return optionsUpdate.call(this);
  }

  public Options inline(String expected)
  {
    return inline(expected, InlineOptions.showCode(true));
  }

  public Options inline(String expected, InlineOptions inlineOptions)
  {
    InlineComparator inline = new InlineComparator(expected, inlineOptions);
    return inline.setForOptions(this);
  }

  public Options(Scrubber scrubber)
  {
    fields.put(Fields.SCRUBBER, scrubber);
  }

  public Options(ApprovalFailureReporter reporter)
  {
    fields.put(Fields.REPORTER, reporter);
  }

  private Options(Map<Fields, Object> fields, Fields key, Object value)
  {
    this.fields.putAll(fields);
    this.fields.put(key, value);
  }

  public ApprovalFailureReporter getReporter()
  {
    return ArrayUtils.getOrElse(fields, Fields.REPORTER, ReporterFactory::get);
  }

  public Options withReporter(ApprovalFailureReporter reporter)
  {
    return new Options(fields, Fields.REPORTER, reporter);
  }

  public Options addReporter(ApprovalFailureReporter reporter)
  {
    return this.withReporter(new MultiReporter(this.getReporter(), reporter));
  }

  public Function2<File, File, VerifyResult> getComparator()
  {
    return ArrayUtils.getOrElse(fields, Fields.COMPARATOR, () -> FileApprover::approveTextFile);
  }

  public Options withComparator(Function2<File, File, VerifyResult> comparator)
  {
    return new Options(fields, Fields.COMPARATOR, comparator);
  }

  public Options withScrubber(Scrubber scrubber)
  {
    return new Options(fields, Fields.SCRUBBER, scrubber);
  }

  public String scrub(String input)
  {
    return getScrubber().scrub(input);
  }

  private Scrubber getScrubber()
  {
    return ArrayUtils.getOrElse(fields, Fields.SCRUBBER, () -> NoOpScrubber.INSTANCE);
  }

  public ApprovalWriter createWriter(Object o)
  {
    ApprovalWriterFactory factory = ArrayUtils.getOrElse(fields, Fields.WRITER,
        DefaultApprovalWriterFactory::getDefaultFactory);
    return factory.create(o, this);
  }

  public Options withWriter(ApprovalWriterFactory approvalWriterFactory)
  {
    return new Options(fields, Fields.WRITER, approvalWriterFactory);
  }

  public FileOptions forFile()
  {
    return new FileOptions(this.fields);
  }
  public static class FileOptions
  {
    private final Map<Fields, Object> fields;
    public FileOptions(Map<Fields, Object> fields)
    {
      this.fields = fields;
    }

    public Options withExtension(String fileExtensionWithDot)
    {
      if (!fileExtensionWithDot.startsWith("."))
      {
        fileExtensionWithDot = "." + fileExtensionWithDot;
      }
      return new Options(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, fileExtensionWithDot);
    }

    public ApprovalNamer getNamer()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_NAMER, Approvals::createApprovalNamer);
    }

    public Options withNamer(ApprovalNamer namer)
    {
      return new Options(fields, Fields.FILE_OPTIONS_NAMER, namer);
    }

    public String getFileExtension()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, () -> ".txt");
    }

    public Options withBaseName(String fileBaseName)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      return withNamer(approvalNamer);
    }

    public Options withName(String fileBaseName, String extension)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      EnumMap<Fields, Object> newFields = new EnumMap<>(fields);
      newFields.put(Fields.FILE_OPTIONS_FILE_EXTENSION, extension);
      return new Options(newFields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }

    public Options withAdditionalInformation(String additionalInformation)
    {
      return withNamer(getNamer().addAdditionalInformation(additionalInformation));
    }
  }
}
