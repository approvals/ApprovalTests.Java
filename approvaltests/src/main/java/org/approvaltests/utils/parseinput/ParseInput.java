package org.approvaltests.utils.parseinput;

import com.spun.util.Tuple;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.lambda.functions.Function1;
import org.lambda.query.Queryable;

import java.util.HashMap;
import java.util.Map;

public class ParseInput<OUT>
{
  private final String                                expected;
  private final ParseInputOptions                     options;
  private final Function1<String, Tuple<String, OUT>> transformer;
  ParseInput(String expected, Function1<String, Tuple<String, OUT>> transformer, ParseInputOptions options)
  {
    this.expected = expected;
    this.transformer = transformer;
    this.options = options;
  }

  public static ParseInput<String> from(String expected)
  {
    return new ParseInput<String>(expected, s -> new Tuple<>(s, s), new ParseInputOptions());
  }

  public static String getLast(Queryable<String> temp, int number)
  {
    Queryable<String> skip = temp.skip(number);
    return skip.size() == 1 ? skip.first() : skip.join(", ");
  }

  public ParseInput<OUT> multiline()
  {
    return new ParseInput<>(expected, transformer, new ParseInputOptions(true));
  }

  public Queryable<Tuple<String, OUT>> parse()
  {
    Function1<String, Boolean> f = options.multiline ? s -> s.contains("->") : s -> true;
    return Queryable.as(expected.split("\n")) //
        .where(f) //
        .select(l -> l.split("->")[0].trim()) //
        .where(l -> !l.isEmpty()) //
        .select(l -> transformer.call(l));
  }

  public Queryable<OUT> getInputs()
  {
    return parse().select(Tuple::getSecond);
  }

  public void verifyAll(Function1<OUT, Object> transform)
  {
    Approvals.verifyAll("", parse(), s -> String.format("%s -> %s", s.getFirst(), transform.call(s.getSecond())),
        new Options().inline(expected));
  }

  public static <OUT> Function1<String, OUT> getTransformerForClass(Class<OUT> targetType)
  {
    Map<Class<?>, Function1<String, Object>> transformers = new HashMap<Class<?>, Function1<String, Object>>()
    {
      {
        put(Integer.class, s -> Integer.parseInt(s));
        put(String.class, s -> s);
        put(Double.class, s -> Double.parseDouble(s));
        put(Boolean.class, s -> Boolean.parseBoolean(s));
        put(Long.class, s -> Long.parseLong(s));
        put(Float.class, s -> Float.parseFloat(s));
        put(Short.class, s -> Short.parseShort(s));
        put(Byte.class, s -> Byte.parseByte(s));
      }
    };
    if (targetType.isArray())
    {
      Class<?> componentType = targetType.getComponentType();
      Function1<String, Object> transformer = transformers.get(componentType);
      return s -> (OUT) Queryable.as(s.split(",")).select(String::trim).select(transformer).asArray();
    }
    else
    {
      return (Function1<String, OUT>) transformers.get(targetType);
    }
  }

  // ************* 1 parameter
  public <T1> ParseInputWith1Parameters<T1> withTypes(Class<T1> type1)
  {
    return ParseInputWith1Parameters.create(expected, type1, options);
  }

  public <T1> ParseInputWith1Parameters<T1> transformTo(Function1<String, T1> transformer)
  {
    return new ParseInputWith1Parameters<>(expected, transformer, options);
  }

  // ************* 2 parameters
  public <T1, T2> ParseInputWith2Parameters<T1, T2> withTypes(Class<T1> type1, Class<T2> type2)
  {
    return ParseInputWith2Parameters.create(expected, getTransformerForClass(type1), getTransformerForClass(type2),
        options);
  }

  public <T1, T2> ParseInputWith2Parameters<T1, T2> transformTo(Function1<String, T1> transformer1,
      Function1<String, T2> transformer2)
  {
    return ParseInputWith2Parameters.create(expected, transformer1, transformer2, options);
  }

  public <T1, T2, T3> ParseInputWith3Parameters<T1, T2, T3> withTypes(Class<T1> type1, Class<T2> type2,
      Class<T3> type3)
  {
    return ParseInputWith3Parameters.create(expected, getTransformerForClass(type1), getTransformerForClass(type2),
        getTransformerForClass(type3), options);
  }

  public <T1, T2, T3> ParseInputWith3Parameters<T1, T2, T3> transformTo(Function1<String, T1> transformer1,
      Function1<String, T2> transformer2, Function1<String, T3> transformer3)
  {
    return ParseInputWith3Parameters.create(expected, transformer1, transformer2, transformer3, options);
  }
  public static class ParseInputOptions
  {
    public final boolean multiline;
    public ParseInputOptions()
    {
      this(false);
    }

    public ParseInputOptions(boolean multiline)
    {
      this.multiline = multiline;
    }
  }
}
