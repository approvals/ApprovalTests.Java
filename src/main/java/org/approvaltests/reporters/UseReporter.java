package org.approvaltests.reporters;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import org.approvaltests.core.ApprovalFailureReporter;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
  Class<? extends ApprovalFailureReporter>[] value();
}
