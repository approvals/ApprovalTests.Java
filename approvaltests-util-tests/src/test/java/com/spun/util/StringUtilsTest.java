package com.spun.util;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.AutoApproveReporter;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.utils.parseinput.ParseInput;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class StringUtilsTest
{
  ReplaceUseCase[] replaceUseCases = {new ReplaceUseCase("quick brown fox", "brown", "white", "quick white fox"),
                                      new ReplaceUseCase("quick brown quick fox brown", "quick", "slow",
                                          "slow brown slow fox brown"),};
  @Test
  public void testJoin()
  {
    assertEquals("Falco,Llewellyn", StringUtils.join("Falco", ",", "Llewellyn"));
    assertEquals("Falco", StringUtils.join("Falco", ",", null));
    assertEquals("Falco", StringUtils.join("Falco", ",", ""));
  }
  @Test
  public void testToNameCase()
  {
    assertEquals("Mr. Frank M Peter", StringUtils.toNameUpperCase("mr. frank m peter"), "Name changed");
  }
  @Test
  public void testSplit()
  {
    var expected = """
        'quick brown fox'.split( ) => [quick, brown, fox]
        'quick/brown/ fox'.split(/) => [quick, brown, fox]
        'quick**brown**fox'.split(\\*\\*) => [quick, brown, fox]
        ' quick   brown fox '.split( ) => [, quick, , , brown, fox]
        'quick brown fox'.split(brown) => [quick, fox]
        """;
    SplitUseCase[] split = {new SplitUseCase("quick brown fox", " "),
                            new SplitUseCase("quick/brown/ fox", "/"),
                            new SplitUseCase("quick**brown**fox", "\\*\\*"),
                            new SplitUseCase(" quick   brown fox ", " "),
                            new SplitUseCase("quick brown fox", "brown"),};
    Approvals.verifyAll(split, a -> String.format("'%s'.split(%s) => %s", a.start, a.splitOn,
        Arrays.toString(StringUtils.split(a.start, a.splitOn, true))), new Options().inline(expected));
  }
  @Test
  public void testJavaScript()
  {
    String[] strings = {"\"\n", "this is a note \"\'\nanother liner"};
    Approvals.verifyAll(strings, a -> multiline(a, StringUtils.toJavaScriptEncode(a)));
  }
  public static String multiline(String from, String to)
  {
    return (from + "\n => \n" + to + "\n" + "------------------------------------");
  }
  @Test
  public void testJoinCollection()
  {
    // begin-snippet: join_collection
    List<Integer> number = Arrays.asList(1, 2, 3, 4, 5);
    String text = StringUtils.join(number, ", ");
    // end-snippet
    Approvals.verify(text);
  }
  @Test
  public void testJoinCollectionWithFunction()
  {
    // begin-snippet: join_collection_with_lambda
    List<Integer> number = Arrays.asList(1, 2, 3, 4, 5);
    String text = StringUtils.join(number, ", ", n -> StringUtils.padNumber(n, 3));
    // end-snippet
    Approvals.verify(text);
  }
  @Test
  public void testReplace()
  {
    for (int i = 0; i < replaceUseCases.length; i++)
    {
      assertEquals(StringUtils.replace(replaceUseCases[i].startingString, replaceUseCases[i].find,
          replaceUseCases[i].replace), replaceUseCases[i].expectedString, "Replace failed");
    }
  }
  @Test
  public void testWriteToString()
  {
    Approvals.verify(StringUtils.toString("things", new Object[]{1, null, "hi"}));
  }
  @Test
  void testSortedMap()
  {
    Map<String, String> sortedMap = new TreeMap<>((s1, s2) -> s1.length() - s2.length());
    sortedMap.put("c", "1");
    sortedMap.put("mmmm", "1");
    sortedMap.put("aaa", "1");
    sortedMap.put("zz", "1");
    Approvals.verify(sortedMap);
  }
  @Test
  void testNonComparableKeysMap()
  {
    Map<Key, String> unsortableMap = new LinkedHashMap<>();
    unsortableMap.put(new Key("d"), "4");
    unsortableMap.put(new Key("a"), "1");
    unsortableMap.put(new Key("c"), "3");
    unsortableMap.put(new Key("b"), "2");
    Approvals.verify(unsortableMap);
  }
  @Test
  void testNonComparableKeysEmptyMap()
  {
    Map<Key, String> unsortableMap = Map.of();
    assertDoesNotThrow(() -> StringUtils.toString(unsortableMap));
  }
  public class Key
  {
    String id;
    public Key(String id)
    {
      this.id = id;
    }
    @Override
    public String toString()
    {
      return id;
    }
    @Override
    public boolean equals(Object o)
    {
      if (!(o instanceof Key))
        return false;
      Key key = (Key) o;
      return Objects.equals(id, key.id);
    }
    @Override
    public int hashCode()
    {
      return Objects.hashCode(id);
    }
  }
  public class SplitUseCase
  {
    String   start;
    String   splitOn;
    String[] expectedArray;
    public SplitUseCase(String startingString, String splitOn, String... expectedArray)
    {
      this.start = startingString;
      this.splitOn = splitOn;
      this.expectedArray = expectedArray;
    }
  }
  public class ReplaceUseCase
  {
    String startingString;
    String find;
    String replace;
    String expectedString;
    public ReplaceUseCase(String startingString, String find, String replace, String expectedString)
    {
      this.startingString = startingString;
      this.find = find;
      this.replace = replace;
      this.expectedString = expectedString;
    }
  }
  @Test
  @UseReporter(AutoApproveReporter.class)
  public void testSplitting()
  {
    var expected = """
        1a2aa3, a -> [1, 2, , 3]
        1ś2śś3, ś -> [1, 2, , 3]
        1a2aa3a, a -> [1, 2, , 3]
        1ś2śś3ś, ś -> [1, 2, , 3]
        1a2aa3aa, a -> [1, 2, , 3, ]
        1ś2śś3śś, ś -> [1, 2, , 3, ]
        """;
    ParseInput.from(expected).withTypes(String.class, String.class)
        .verifyAll((i, p) -> Arrays.toString(StringUtils.split(i, p)));
  }
  @Test
  @UseReporter(AutoApproveReporter.class)
  public void testRepeating()
  {
    var expected = """
        Hello, 3 -> HelloHelloHello
        -, 56 -> --------------------------------------------------------
        """;
    ParseInput.from(expected).withTypes(String.class, Integer.class).verifyAll((s, i) -> StringUtils.repeat(s, i));
  }
  @Test
  public void testEnsureEnding()
  {
    String first = "hello";
    String second = "hello\n";
    assertEquals(StringUtils.ensureEnding(first, "\n"), second);
    assertEquals(StringUtils.ensureEnding(second, "\n"), second);
  }
}