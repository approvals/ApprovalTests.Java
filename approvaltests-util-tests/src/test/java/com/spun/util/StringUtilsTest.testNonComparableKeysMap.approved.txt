d : 4 
a : 1 
c : 3 
b : 2 
